### ISSUE TITLE
kolibri never shuts down cleanly

### ISSUE THREAD

**Issue by intelliant01** [learningequality/kolibri-server#20]
Every shutdown attempt either with ```systemctl stop kolibri-server``` or ```service stop kolibri-server``` fails.

Then an attempt with ```systemctl stop kolibri``` or ```service kolibri stop``` seems to kill it in the background with these in the log -

```
ERROR 2019-07-30 18:45:45,443 cli During graceful shutdown, server says: Failed to start (check log file: /home/mssadmin/.kolibri/logs/daemon.txt) (6)
ERROR 2019-07-30 18:45:45,623 cli Not responding, killing with force
```

Contents of ```daemon.txt``` -

```
INFO     Attempting a ping.
INFO     Started new JOBCHECKER thread ID#5f108dffae1f4cdab6b90a1dec235b3c
INFO     Updating num_coach_content on existing channels
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/kolibri/dist/sqlalchemy/engine/base.py", line 1193, in _execute_context
    context)
  File "/usr/lib/python3/dist-packages/kolibri/dist/sqlalchemy/engine/default.py", line 509, in do_execute
    cursor.execute(statement, parameters)
sqlite3.OperationalError: database is locked

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/bin/kolibri", line 9, in <module>
    load_entry_point('kolibri==0.12.7', 'console_scripts', 'kolibri')()
  File "/usr/lib/python3/dist-packages/kolibri/utils/cli.py", line 625, in main
    start(port, daemon=daemon)
  File "/usr/lib/python3/dist-packages/kolibri/utils/cli.py", line 319, in start
    server.start(port=port, run_cherrypy=run_cherrypy)
  File "/usr/lib/python3/dist-packages/kolibri/utils/server.py", line 113, in start
    update_channel_metadata()
  File "/usr/lib/python3/dist-packages/kolibri/core/content/utils/annotation.py", line 73, in update_channel_metadata
    update_num_coach_contents()
  File "/usr/lib/python3/dist-packages/kolibri/core/content/utils/annotation.py", line 167, in update_num_coach_contents
    .values(num_coach_contents=cast(ContentNodeTable.c.coach_content, Integer()))
  File "/usr/lib/python3/dist-packages/kolibri/dist/sqlalchemy/engine/base.py", line 948, in execute
    return meth(self, multiparams, params)
  File "/usr/lib/python3/dist-packages/kolibri/dist/sqlalchemy/sql/elements.py", line 269, in _execute_on_connection
    return connection._execute_clauseelement(self, multiparams, params)
  File "/usr/lib/python3/dist-packages/kolibri/dist/sqlalchemy/engine/base.py", line 1060, in _execute_clauseelement
    compiled_sql, distilled_params
  File "/usr/lib/python3/dist-packages/kolibri/dist/sqlalchemy/engine/base.py", line 1200, in _execute_context
    context)
  File "/usr/lib/python3/dist-packages/kolibri/dist/sqlalchemy/engine/base.py", line 1413, in _handle_dbapi_exception
    exc_info
  File "/usr/lib/python3/dist-packages/kolibri/dist/sqlalchemy/util/compat.py", line 265, in raise_from_cause
    reraise(type(exception), exception, tb=exc_tb, cause=cause)
  File "/usr/lib/python3/dist-packages/kolibri/dist/sqlalchemy/util/compat.py", line 248, in reraise
    raise value.with_traceback(tb)
  File "/usr/lib/python3/dist-packages/kolibri/dist/sqlalchemy/engine/base.py", line 1193, in _execute_context
    context)
  File "/usr/lib/python3/dist-packages/kolibri/dist/sqlalchemy/engine/default.py", line 509, in do_execute
    cursor.execute(statement, parameters)
sqlalchemy.exc.OperationalError: (sqlite3.OperationalError) database is locked [SQL: 'UPDATE content_contentnode SET num_coach_contents=CAST(content_contentnode.coach_content AS INTEGER) WHERE content_contentnode.kind != ?'] [parameters: ('topic',)] (Background on this error at: http://sqlalche.me/e/e3q8)
INFO     Sqlite database Vacuum finished.
INFO     Next Vacuum at 3AM local server time (in 519.7333333333333 minutes).
INFO     Ping succeeded! (response: {'id': 294786, 'messages': []})
INFO     Sleeping for 1440.0 minutes.
```

Residual db.sqlite3-shm and db.sqlite3-wal files are always present, guess indicative of an unclean shutdown.

Could be related to #19 as I see a database lock error.

kolibri-server 0.3.0
kolibri 0.12.7
Ubuntu 18.04

**Comment by benjaoming** [comment]
This is probably because kolibri is active in UWSGI.

But does seem a bit critical.

@intelliant01 could you clarify the use case of performing this `stop`? Was it done automatically or as part of a task you needed to solve?

**Comment by intelliant01** [comment]
@benjaoming I was testing the normal start / stop cycle post installation of kolibri-server package. When I spotted the odd behaviour with the above errors , I tried this multiple times and each time I spotted the same error. I shall be doing a round of tests on a clean install tomorrow.

I say *clean* because the above is a case of migrating from pex to just deb to kolibri-server deb. Though I had taken due care to cleanup before each transition but yet want to be sure with a kolibri-server only installation.

**Comment by intelliant01** [comment]
@benjaoming all works smooth on a new OS install followed by a kolibri-server installation. The problem is specific to the pex to kolibri-server migrated system.
Will now focus on figuring out what's happening therein.

**Comment by intelliant01** [comment]
@benjaoming Explaining two scenarios:

- Immediately after kolibri-server installation / small db.sqlite3 file -

When stopping with ```systemctl stop kolibri-server``` uwsgi processes are terminated but ```/usr/bin/python3 /usr/bin/kolibri start``` process can be seen running as well as the ```systemctl status kolibri``` is also in active state. When stopped with ```systemctl stop kolibri```, it reports a clean shutdown -

```
# systemctl status kolibri
● kolibri.service - LSB: kolibri daemon, an offline education platform
   Loaded: loaded (/etc/init.d/kolibri; disabled; vendor preset: enabled)
   Active: inactive (dead) since Thu 2019-08-08 01:38:48 IST; 35s ago
  Process: 9572 ExecStop=/etc/init.d/kolibri stop (code=exited, status=0/SUCCESS)
  Process: 3831 ExecStart=/etc/init.d/kolibri start (code=exited, status=0/SUCCESS)

Aug 07 19:58:06 server kolibri[3831]: INFO     Starting Kolibri background services
Aug 07 19:58:06 server kolibri[3831]: INFO     Going to daemon mode, logging to /home/mssadmin/.kolibri/logs/kolibri.txt
Aug 07 19:58:07 server systemd[1]: Started LSB: kolibri daemon, an offline education platform.
Aug 08 01:38:44 server systemd[1]: Stopping LSB: kolibri daemon, an offline education platform...
Aug 08 01:38:44 server su[9597]: Successful su for mssadmin by root
Aug 08 01:38:44 server su[9597]: + ??? root:mssadmin
Aug 08 01:38:44 server su[9597]: pam_unix(su:session): session opened for user mssadmin by (uid=0)
Aug 08 01:38:47 server kolibri[9572]: INFO     Running Kolibri with the following settings: kolibri.deployment.default.settings.base
Aug 08 01:38:47 server kolibri[9572]: INFO     Kolibri background services have successfully been stopped.
Aug 08 01:38:48 server systemd[1]: Stopped LSB: kolibri daemon, an offline education platform.
```

but ```db.sqlite3-wal``` and ```db.sqlite3-shm``` residual files always remain.

- With a large database (approx 2.7 GB), uwsgi processes are terminated properly (verified by ```ps aux | grep kolibri```) with ```systemctl stop kolibri-server``` but ```systemctl status kolibri``` returns active. 
Thereafter when terminated with ```service kolibri stop``` (as is being done in /etc/init.d/kolibri-server) deliberately or ```systemctl stop kolibri```, I always get this unclean shutdown -

```
# systemctl status kolibri
● kolibri.service - LSB: kolibri daemon, an offline education platform
   Loaded: loaded (/etc/init.d/kolibri; disabled; vendor preset: enabled)
   Active: inactive (dead)

Aug 08 01:25:59 server kolibri[31310]: INFO     Starting Kolibri background services
Aug 08 01:25:59 server kolibri[31310]: INFO     Going to daemon mode, logging to /home/mssadmin/.kolibri/logs/kolibri.txt
Aug 08 01:26:00 server systemd[1]: Started LSB: kolibri daemon, an offline education platform.
Aug 08 01:48:12 server systemd[1]: Stopping LSB: kolibri daemon, an offline education platform...
Aug 08 01:48:12 server su[1811]: Successful su for mssadmin by root
Aug 08 01:48:12 server su[1811]: + ??? root:mssadmin
Aug 08 01:48:15 server kolibri[1781]: INFO     Running Kolibri with the following settings: kolibri.deployment.default.settings.base
Aug 08 01:48:16 server kolibri[1781]: ERROR    During graceful shutdown, server says: Failed to start (check log file: /home/mssadmin/.kolibri/logs/daemon.txt) (6)
Aug 08 01:48:16 server kolibri[1781]: ERROR    Not responding, killing with force
Aug 08 01:48:17 server systemd[1]: Stopped LSB: kolibri daemon, an offline education platform.

```
In this case also ```db.sqlite3-wal``` and ```db.sqlite3-shm``` residual files always remain.

For you to replicate this, would you like access to the large database file?

**Comment by jredrejo** [comment]
@intelliant01 as the problem comes because of transitioning from running pex and can't be reproduced in a fresh kolibri-server installation I only can think of two possible reasons:
1. The database was already damaged and vacuum can't finish its job
2. There are still some remaining script in the server keeping the database open

For 1, it could be easily solved using this procedure: https://kolibri.readthedocs.io/en/latest/manage.html?highlight=database#troubleshoot-database-issues

For 2 , it's a bit hard to troubleshoot, could you execute (in the directory where the db is) ?
`lsof job_storage.sqlite3`

If the output of this command show lines like:
```
kolibri 10910 jose   15ur  REG    8,1    45056 536525 job_storage.sqlite3
uwsgi   11070 jose   10ur  REG    8,1    45056 536525 job_storage.sqlite3
```
we could check the PID of the process keeping the db open (the pid is the second number in each row).

If we can't make any progress with these steps, would it be possible you share the db somehow? (via dropbox, or a zipped file in some shared gdrive folder, etc.) You can send me a dm with the link if you prefer not to share it in public.

**Comment by intelliant01** [comment]
@jredrejo This problem is easily reproducible on a clean kolibri-server installation i.e. without the pex to deb/kolibri-server transition. 

Have tried with troubleshooting steps, manual vaccumming, lsof, ps aux checks, etc. You shall easily be able to reproduce the shutdown and startup problems once you play with the large database on a dual core / slow system i.e. if you have access to one. :)

Will shortly share the gdrive link to the compressed database.

**Comment by intelliant01** [comment]
@jredrejo [Herein is the link to the db](https://drive.google.com/file/d/1JdqS8E_hOTt_-7BWFS-AZQpbRV_3dF0e/view?usp=sharing). Shall give access once you request.

**Comment by jredrejo** [comment]
@intelliant01 Thanks very much for your feedback. I have been able to reproduce the problem now.
There were 2 problems:
1. A problem with kolibri-server init script that didn't wait for `service kolibri stop`to finish. This is being fixed in #28 . A new version of kolibri-server is going to be released today to fix it.
2. A problem with one function that's been executed whenever kolibri starts and keeps the db open while vacuum is done, thus vacuum does not work properly. This is only reproducible in big databases as the one you had. This problem has been fixed in kolibri ( in https://github.com/learningequality/kolibri/pull/5494 ) but the fix is not included in the latest release, it will be included in the 0.13.x version.

So, for 2 you will need to wait for the next kolibri version to be released. Sorry for that.

**Comment by intelliant01** [comment]
@jredrejo Thanks for the heads up. 
Will test for 1 in my environment once released. Really waiting for fix referenced in 2 as the bug results in all sorts of inconsistent behaviour for end users. What could be the approx ETA for the next 0.13.x release?

**Comment by jredrejo** [comment]
> What could be the approx ETA for the next 0.13.x release?

I am not sure, I think @indirectlylit can answer it properly.

**Comment by indirectlylit** [comment]
We don't have a hard date for 0.13.x, but it will be before the end of the year.

I've created #5891 to investigate whether we can backport the fix for (2) to 0.12.x.

**Comment by intelliant01** [comment]
@indirectlylit Thanks for the heads up. Really hoping that it can be bavkported as waiting till year end will just prolong the agony of inconsistent experiences for many a on-field users.
